src/main/java/ch/njol/skript/expressions/ExprName.java

```diff
@@ -97,336 +96,151 @@
+			return m.hasDisplayName() ? m.getDisplayName() : null;
+		} else if (o instanceof Inventory) {
+			if (TITLE_METHOD != null) {
+				try {
+					return (String) TITLE_METHOD.invoke(o);
+				} catch (Throwable e) {
+					Skript.exception(e);
+					return null;
+				}
+			} else {
+				if (!((Inventory) o).getViewers().isEmpty())
+					return ((Inventory) o).getViewers().get(0).getOpenInventory().getTitle();
+				return null;
+			}
+		} else if (o instanceof Slot) {
+			ItemStack is = ((Slot) o).getItem();
+			if (is != null && is.hasItemMeta()) {
+				ItemMeta m = is.getItemMeta();
+				return m.hasDisplayName() ? m.getDisplayName() : null;
+			}
+		}
+		return null;
 	}
 
-	private int changeType = 0;
-
-	// TODO find a better method for handling changes (in general)
-	// e.g. a Changer that takes an object and returns another which should then be saved if applicable (the Changer includes the ChangeMode)
 	@Override
 	@Nullable
 	public Class<?>[] acceptChange(ChangeMode mode) {
-		if (mode == ChangeMode.DELETE && (type.acceptChange & ~PLAYER) != 0 || mode == ChangeMode.RESET)
-			return new Class[0];
-		if (mode != ChangeMode.SET)
-			return null;
-		if ((type.acceptChange & PLAYER) != 0 && Player.class.isAssignableFrom(getExpr().getReturnType())) {
-			changeType = PLAYER;
-		} else if ((type.acceptChange & INVENTORY) != 0 && Inventory.class.isAssignableFrom(getExpr().getReturnType())) {
-			changeType = INVENTORY;
-		} else if ((type.acceptChange & ITEM) != 0 && (getExpr().isSingle() && ChangerUtils.acceptsChange(getExpr(), ChangeMode.SET, ItemStack.class, ItemType.class) || Slot.class.isAssignableFrom(getExpr().getReturnType()))) {
-			changeType = ITEM;
-		} else if ((type.acceptChange & ENTITY) != 0 && Entity.class.isAssignableFrom(getExpr().getReturnType())) {
-			if (type == NameType.NAME && Player.class.isAssignableFrom(getExpr().getReturnType())) {
+		if (mode == ChangeMode.SET || mode == ChangeMode.RESET) {
+			if (mark == 1 && Player.class.isAssignableFrom(getExpr().getReturnType())) {
 				Skript.error("Can't change the Minecraft name of a player. Change the 'display name' or 'tab list name' instead.");
 				return null;
 			}
-			changeType = ENTITY;
+			return CollectionUtils.array(String.class);
 		}
-		return changeType == 0 ? null : CollectionUtils.array(String.class);
+		return null;	
 	}
 
 	@Override
 	public void change(Event e, @Nullable Object[] delta, ChangeMode mode) {
-		String name = delta == null ? null : (String) delta[0];
-		if (changeType == ITEM) {
-			if (Slot.class.isAssignableFrom(getExpr().getReturnType())) {
-				for (Slot s : (Slot[]) getExpr().getArray(e)) {
-					ItemStack i = s.getItem();
-					type.set(i, name);
-					s.setItem(i);
+		String name = delta != null ? (String) delta[0] : null;
+		for (Object o : getExpr().getArray(e)) {
+			if (o instanceof Player) {
+				switch (mark) {
+					case 2: ((Player) o).setDisplayName(name != null ? name + ChatColor.RESET : ((Player) o).getName());

```

Comment:
missing break statement (even if the following code is unreachable, the compiler throws a warning for it so)